OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyOptionsLogic.cs

```diff
@@ -75,18 +78,48 @@ void RebuildOptions()
 
 			optionsContainer.RemoveChildren();
 			optionsContainer.Bounds.Height = 0;
+
+			var categories = mapPreview.WorldActorInfo.TraitInfos<LobbyOptionCategoryInfo>()
+				.OrderBy(o => o.DisplayOrder)
+				.ToArray();
+
 			var allOptions = mapPreview.PlayerActorInfo.TraitInfos<ILobbyOptions>()
 					.Concat(mapPreview.WorldActorInfo.TraitInfos<ILobbyOptions>())
 					.SelectMany(t => t.LobbyOptions(mapPreview))
 					.Where(o => o.IsVisible)
 					.OrderBy(o => o.DisplayOrder)
 					.ToArray();
 
+			RebuildCategory(allOptions.Where(o => o.Category == null || categories.All(c => c.Category != o.Category)).ToArray());
+
+			foreach (var category in categories)
+			{
+				var headline = (LabelWidget)headlineTemplate.Clone();
+				headline.GetText = () => category.Title;
+				headline.Bounds.Y = optionsContainer.Bounds.Height;
+
+				optionsContainer.Bounds.Height += headline.Bounds.Height;
+				optionsContainer.AddChild(headline);
+
+				RebuildCategory(allOptions.Where(o => o.Category == category.Category).ToArray());
+			}
+
+			panel.ContentHeight = yMargin + optionsContainer.Bounds.Height;
+			optionsContainer.Bounds.Y = yMargin;
+
+			if (panel.ContentHeight < panel.Bounds.Height)
+				optionsContainer.Bounds.Y += (panel.Bounds.Height - panel.ContentHeight) / 2;
+
+			panel.ScrollToTop();
+		}
+
+		void RebuildCategory(LobbyOption[] options)
```

Comment:
Is there a reason this is an array, or can we change this to `IEnumerable<LobbyOption>` and skip the `ToArray` calls above?

Reply:
Rider nags about "Possible multiple enumeration". So theoretically this method may get a not-multiple-times-enumerable value. But sure, i can make it an ienumerable ;)

Reply:
Goes to show that you shouldn't always trust Rider. Unless I'm missing something, `RebuildCategory` should be changed to have one for loop akin to this:
```
			foreach (var option in options)
			{
				if (option is LobbyBooleanOption)
				{
					[... block that was inside the first foreach ...]
				}
				else
				{
					[... block that was inside the second foreach ...]
				}
			}
```
Then we only enumerate once.

Reply:
merged the loops, its now an IEnumerable. Well actualy the info from rider was right -> it only shows that warning when looping an ienumerable twice, which is what did actualy happen before.